`Discussion`: Improve announcement post handling and formatting

lschlesinger opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR addresses several shortcomings of the current implementation and enhances some announcement/notification related features:
- Answer post notification in the notification side-bar showed author of original post as "author" and not the answer post that is notified about. This was misleading.
- Info banner on announcement creation is improved: instructor is informed about automatic email notification
- Email notification on announcement changed format -> title and course are in subject (Betreff) instead of "New announcement".
- Content in email was not shown as html rendered markdown but rather raw markdown -> formatting of original post content was lost in email, now it's the same format.
- Code block and reference were not rendered correctly anymore in post content due to a markdown issue introduced earlier
- Tutors could edit an announcement, but not create -> now they are  _not_ allowed to edit or create an announcement

### Description
<!-- Describe your changes in detail -->
I addressed the shortcomings/enhancements mentioned above. Most of same imply just minor code changes.
However, I had to add a new pipe `htmlForPostingMarkdown` that mainly relies on the same logic behind the existing `htmlForMarkdown` pipe.
For the markdown used in posting content we are facing an additional challenge which are the integrated post references.
To enable rendering the posting content including a link in between, we split up the content into the parts: `contentBeforeReference`, `reference` and `contentAfterReference`. Each of them can contain multiple paragraphs after converting markdown to html.
In order to assure that the reference is in the same line as the text it is embedded we need to manipulate the paragraph that is directly before the reference and the paragraph that comes directly after the reference. Those should have the css attribute `display: inline` explicitly set.
This was previously achieved in a different way which broke the rendering of code-block markdown and reference markdown.
Now, with the dedicated pipe `htmlForPostingContent` we achieve that the markdown formatting is not broken for code-block and reference _and_ the reference (if there is one) is in the same line as the content around it.

Furthermore, we added the lightweight markdown-to-html paser **[commonmark](https://github.com/commonmark/commonmark-java)** as new server dependency.
Now, we are able to parse the post content (markdown) into html that can be used within the email template on the server-side.
We decided to do this on the server-side as we cannot guarantee security if the html content that is directly used as mail content can be created (and manipulate by users though intercepting the request!) on client-side.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 1 Tutor
- 1 Course with postings enabled

**TEST ON TS1 to receive email notifications**

1. Log in to Artemis
2. Navigate to Course Discussion
3. As Tutor, create an answer post to an existing post (that was created by someone else) -> check the notification, the name shown should be the author of the answer post (not the post)
4. As tutor, find an announcement in the course posts -> check that there is not icon to edit or delete it
5. As tutor, create a new post on the course overview -> check that you cannot select `Announcement` as topic
6. As tutor, find an existing post, that is no announcement post, edit it -> check that you cannot select `Announcement` as topic
7. As instructor, start create a new announcement post (create new post at course overview, select `Announcement` as topic) -> check that the blue info box contains the information on email notification
8. As content use markdown formatting that is supported by the editor (bold, italic, underline, code, codeblock, reference, link) and including a reference (e.g #ID)
9. Save the announcement -> see if the content is rendered correctly
10. Check the email that was sent -> check if it has the title of the announcement as well as the course as subject, -> check if the content of the mail is formatted as seen on the artemis page

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->
posting-header: 100%
markdown.service.ts: 83.6% (newly added part completely covered by `postings-content-part.component.spec.ts`)
post-header.component.ts: 100%
posting-content-part.component.ts: 100%
html-for-post-markdown.pipe.ts: 100%

Applied Changes to GroupNotificationServiceTest (still ~100%)

Local tests all pass
![image](https://user-images.githubusercontent.com/65814168/142433668-7e712513-6817-4b38-a1bb-bf664887c860.png)


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

Email Content Before:
<img width="1206" alt="Bildschirmfoto 2021-11-18 um 14 49 45" src="https://user-images.githubusercontent.com/41366522/142427609-6ad3335d-f3dc-47f5-8cdd-f7fa9e99c888.png">

Email Content After:
![image](https://user-images.githubusercontent.com/41366522/142599974-9ffc6ed0-e606-4e2c-9af5-9653f6631956.png)


Markdown Post Content Before:

<img width="703" alt="Bildschirmfoto 2021-11-18 um 14 48 03" src="https://user-images.githubusercontent.com/41366522/142427529-12687ec1-33c5-41a6-8cfe-3f333e974cb5.png">

Markdown Post Content After:
<img width="1176" alt="Bildschirmfoto 2021-11-18 um 14 53 47" src="https://user-images.githubusercontent.com/41366522/142428513-b40e57d3-7ef3-49d6-b64c-cab8db509c92.png">

lschlesinger commented on src/main/resources/templates/mail/notification/announcementPostEmail.html:
```diff
@@ -7,14 +7,7 @@
-        <!-- Change message : notification text -->
-        <p class="notification-text-header" th:text="#{email.notification.aux.notification.post.content}"/>
-        <p class="notification-text" th:text="${notificationSubject.getContent()}">Post Content</p>
+        <p th:utext="${notificationSubject.getContent()}">Post Content</p>
```
@Malyuk-A I am not sure how I should integrate strings including html tags here...is `utext` (unescaped text) the correct approach?

MDK5 reviewed (changes requested):
Tested in testings session, two issues found:

- As mentioned, the endpoints should be also secured to defend against request forgery
- The URLs in the emails are broken/not existent

lschlesinger commented:
@MDK5 I fixed it, as can be seen in the updated screenshot of the email content. The links are displayed now and work as expected.
Also I added additional server-side checks for preventing tutors from creating, editing or deleting an announcement. I also added according test cases.

Olimbia23 commented on src/main/java/de/tum/in/www1/artemis/service/MailService.java:
```diff
@@ -153,6 +153,48 @@ public void sendSAML2SetPasswordMail(User user) {
+     * @param context that is modified based on the exercise type
+     */
+    private void setExerciseContext(Object notificationSubject, Context context) {
+        if (notificationSubject instanceof QuizExercise) {
+            context.setVariable(EXERCISE_TYPE, "quiz");
+        }
+        else if (notificationSubject instanceof ModelingExercise) {
+            context.setVariable(EXERCISE_TYPE, "modeling");
+        }
+        else if (notificationSubject instanceof TextExercise) {
+            context.setVariable(EXERCISE_TYPE, "text");
+        }
+        else if (notificationSubject instanceof ProgrammingExercise) {
+            context.setVariable(EXERCISE_TYPE, "programming");
+        }
+        else if (notificationSubject instanceof FileUploadExercise) {
+            context.setVariable(EXERCISE_TYPE, "upload");
+        }
```
[nit] What do you think of introducing `getExerciseType()` method on the Exercise class and then provide the trivial implementation for it, as in:
```java
enum ExerciseType {
  case TEXT, MODELING, ...
}

class TextExercise {
  ExerciseType getExerciseType() {
    return ExerciseType.TEXT;
  }
}
```

We can make this method (`getExerciseType`) as required one and then if new Exercise is introduced, this code logic (`setExerciseContext`) will just work without any extra though.

I guess otherwise it makes sense to at least log (or assert) in the `else` case, as we won't set the exercise for the exercise which will hit the else case.

PS. Oh, just noticed it's not your code.

lschlesinger replied:
yes, same applies to the other comments. I think it's better to address this in a mail/notification related PR. But thanks for the input :)

Malyuk-A replied:
Very good idea. I simply crafted something more or less on the fly. I added your change to a follow-up PR (see my answer to your other comment about the MailService for more info)

Olimbia23 commented on src/main/java/de/tum/in/www1/artemis/service/MailService.java:
```diff
@@ -172,35 +214,23 @@ public void sendNotificationEmail(Notification notification, User user, Object n
             // posts use a different mechanism for the url
             context.setVariable(NOTIFICATION_URL, NotificationTarget.extractNotificationUrl((Post) notificationSubject, artemisServerUrl.toString()));
+            subject = setPostContextAndSubject(context, notificationSubject, locale);
+
```
[nit] extra newline

Olimbia23 commented on src/main/java/de/tum/in/www1/artemis/service/MailService.java:
```diff
@@ -153,6 +153,48 @@ public void sendSAML2SetPasswordMail(User user) {
+        // posts use a different mechanism for the url
+        context.setVariable(NOTIFICATION_URL, NotificationTarget.extractNotificationUrl((Post) notificationSubject, artemisServerUrl.toString()));
+
+        if (locale.toString().equals("en")) {
+            return "New announcement \"" + ((Post) notificationSubject).getTitle() + "\" in course \"" + ((Post) notificationSubject).getCourse().getTitle() + "\"";
+        }
+        else {
+            return "Neue Ankündigung \"" + ((Post) notificationSubject).getTitle() + "\" im Kurs \"" + ((Post) notificationSubject).getCourse().getTitle() + "\"";
+        }
```
[nit] I assume there is no translation-like service on the server, as we have on the client, right?
I think we could use formatted strings instead, in order to make it more readable.
Something like:
```
final String newAnnouncementEN = "New announcement \"%s\" in course \"%s\"".
final String newAnnouncementDE = "...";

...
private String setPostContextAndSubject(Context context, Object notificationSubject, Locale locale) {
  ...
  String newAnnouncement = locale.toString().equals("en") ? newAnnouncementEN : newAnnouncementDE ;
  return String.format(newAnnoucement, 
                       ((Post) notificationSubject).getTitle(), 
                       ((Post) notificationSubject).getCourse().getTitle());
}
```
I guess this would also be more easier to port to the translation-like service in the future.

Malyuk-A replied:
This is quite interesting, I have not yet seen this in Java (it reminds me of C). Thank you very much for this idea.
Because this is an issue of code quality and not related to Lorena's work (explicitly) I will implement this (and your other remark) in a small follow-up PR. The logic works and I think this should not hinder this Metis PR from being merged.

Malyuk-A replied:
This is the follow up PR which I will open once this PR has been merged. https://github.com/ls1intum/Artemis/pull/4345

Olimbia23 commented on src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:
```diff
@@ -438,4 +464,18 @@ public Post findById(Long postId) {
+     * @param user      requesting user
+     * @param course    course the posting belongs to
+     */
+    private void mayInteractWithPost(Post post, User user, Course course) {
```
Should we also add the "ElseThrow" suffix to the method name?

lschlesinger replied:
yes, it's more consistent, I will change it.

Olimbia23 reviewed (approved):
Nice work @lschlesinger 
Code changes look good.
Approve 👍 

PS. I left some minor comments

MerlinMehmed reviewed (dismissed):
Tested on TS1, all test steps worked as described!

Olimbia23 reviewed (approved):
Thank you for applying my suggestions @lschlesinger 
Code changes look good.
Approve 👍

Malyuk-A reviewed (dismissed):
Re-approve (minor changes after last review)

willergeorg reviewed (approved):
Tested on TS1. No error occurred and link seems to be in the email again.

MerlinMehmed reviewed (approved):
Reapprove!

Olimbia23 reviewed (approved):
Reapprove 👍
